### DISCUSSION TITLE
Using this as a basis for a theme to use with our apps

### DISCUSSION THREAD

**Discussion by misilot** [helloflask/bootstrap-flask#274] [Q&A]
Hi!

I am trying to use project as the basis for our flask projects, so I can place customizations in one spot, and can reuse them with all of our projects.

```python
# import warnings
from flask_bootstrap import Bootstrap4
from flask import current_app, Markup, Blueprint, url_for
from flask import Flask



class KSULBootstrap4(Bootstrap4):
    """
    Extension class for Bootstrap 4.
    Initialize the extension::
        from flask import Flask
        from flask_bootstrap import Bootstrap4
        app = Flask(__name__)
        bootstrap = Bootstrap4(app)
    Or with the application factory::
        from flask import Flask
        from flask_bootstrap import Bootstrap4
        bootstrap = Bootstrap4()
        def create_app():
            app = Flask(__name__)
            bootstrap.init_app(app)
    .. versionchanged:: 2.0.0
       Move common logic to base class ``_Bootstrap``.
    """
    # bootstrap_version = '4.6.2'
    # jquery_version = '3.5.1'
    # popper_version = '1.16.1'
    # bootstrap_css_integrity = 'sha384-xOolHFLEh07PJGoPkLv1IbcEPTNtaed2xpHsD9ESMhqIYd0nLMwNLD69Npy4HI+N'
    # bootstrap_js_integrity = 'sha384-+sLIOodYLS7CIrQpBjl+C7nPvqq+FbNUBDunl/OZv93DB7Ln/533i8e/mZXLi/P+'
    # jquery_integrity = 'sha384-DfXdz2htPH0lsSSs5nCTpuj/zy4C+OGpamoFVy38MVBnE+IbbVYUew+OrCXaRkfj'
    # popper_integrity = 'sha384-9/reFTGAW83EW2RDu2S0VKaIzap3H66lZH81PoYlFhbGU+6BZp6G7niu735Sk7lN'
    # popper_name = 'popper.js'
    static_folder = 'bootstrap4'


    def __init__(self, app=None):

        # if not hasattr(app, 'extensions'):
            # app.extensions = {}
        # app.extensions['ksul-bootstrap'] = self

        # print(self.static_folder)

        blueprint = Blueprint('ksul-bootstrap', __name__, static_folder=f'static/{self.static_folder}',
                              static_url_path=f'/ksul{app.static_url_path}',
                              template_folder='templates')
        
        # blueprint.add_url_rule('/favicon.ico',
        #     redirect_to=url_for('static', filename='favicon.ico'))

        app.register_blueprint(blueprint)

        super().__init__(app)
```

This seems to work but I am trying to add the /favicon.ico path above however I get the following error 

```
RuntimeError: Working outside of application context.

This typically means that you attempted to use functionality that needed
the current application. To solve this, set up an application context
with app.app_context(). See the documentation for more information.
```

I'm not sure what I am missing.
Thanks!

**Comment by PanderMusubi** [comment]
Favicon can be served if you put it in the static dir and refer there. Does that fix it?

**Comment by misilot** [comment]
So, I could refer to the favicon in the html and using a <link> tag, but I am also trying to have it respond at /favicon.ico which is where some applications / browsers look for the icon without reading the html source. 

I'd also like to build in some specific pages that are always there  in my application when I extend this extension and use it locally.

**Reply by PanderMusubi** [reply]
You can have /favicon.ico too in your deployment in a production webserver, not flask. It will be served and available but outside of your application context.

Your application templates can use it but that won't work in your developement environment with flask.

**Reply by misilot** [reply]
I'm trying to have it served / included in my  application context if that makes sense. Doing it that way so I don't have to place the file on each system, and configure apache/nginx or something else to specially serve that file. 

But the same question applies if I wanted to say add a route to say "/about" that has a base template in my extended class `KSULBootstrap4` that I can overwrite in my application itself for the content, but the route always exists.

**Comment by PanderMusubi** [comment]
Static files and template paths are two different things. But yeah, you could deploy favicon.ico in both / and /static.

**Comment by PanderMusubi** [comment]
According to documentation of Flask, you can set static_folder to an empty string in the Flask constructor to serve static files from / but I'm not sure if that interferes with how Bootstrap-Flask works. The use of mightblueprint  work around it.

See https://github.com/helloflask/bootstrap-flask/blob/master/flask_bootstrap/__init__.py#L62 But please try it and let us know.
